### PR TITLE
Proximity signed double fix

### PR DESCRIPTION
**Pull request by yhahn**
### The Bug

```diff
inline double tileDist(unsigned px, unsigned py, unsigned tileX, unsigned tileY) {
-    const double dx = static_cast<double>(px - tileX);
-    const double dy = static_cast<double>(py - tileY);
-    const double distance = dx * dx + dy * dy;
+    const double dx = static_cast<double>(px) - static_cast<double>(tileX);
+    const double dy = static_cast<double>(py) - static_cast<double>(tileY);
+    const double distance = sqrt((dx * dx) + (dy * dy));
```

What we're looking at:

- px, py are the x,y tile coordinates of the proximity center
- tileX, tileY are the x,y tile coordinates of the cover getting its scoredist calculated for it
- when tileX > px or tileY > py the calculated `dx` or `dy` values will be negative
- but all these values are of type `unsigned` so the result overflows before it's converted to a (signed) `double`
- leads to huge wacky numbers being plugged into the distance equation

The fix:

- cast everything to signed types before doing the operation.

### The Feature

Adds a `radius` (in miles) param to coalesce that lets us control the radius effect of proximity from JS. As we battle test proximity further with more use cases, this will let us keep carmen/carmen-cache's concept of scoredist radius in sync.

I also replaced the hardcoded zoom scale in scoredist with some math.

cc @mapbox/geocoding-gang

### CODE REVIEW

File: src/binding.cpp [hunk 1]
```diff
@@ -1268,26 +1268,31 @@ struct CoalesceBaton : carmen::noncopyable {
...
-    if (zoom == 9)  scoredist = 6.0 / distance;
-    if (zoom == 8)  scoredist = 3.5 / distance;
-    if (zoom == 7)  scoredist = 2.0 / distance;
-    if (zoom <= 6)  scoredist = 1.125 / distance;
+
+    // Since distance is in tiles we calculate scoredist by converting the miles into
+    // a tile unit value at the appropriate zoom first.
+    //
+    // 32 tiles is about 40 miles at z14, use this as our mile <=> tile conversion.
+    scoredist = ((radius*(32.0/40.0)) / _pow(1.5, 14 - zoom)) / distance;
```
Comment (yhahn) [inline src/binding.cpp]: For now this recreates a scale similar to what we had before. In actuality I believe the power here should just be `2^14-z` to keep the radius distance constant across zoom levels.

File: src/binding.cpp [hunk 2]
```diff
@@ -1874,6 +1879,20 @@ NAN_METHOD(coalesce) {
...
+            if (!prop_val->IsNumber()) {
+                return Nan::ThrowTypeError("radius must be a number");
+            }
+            int64_t _radius = prop_val->IntegerValue();
+            if (_radius < 0 || _radius > std::numeric_limits<unsigned>::max()) {
+                return Nan::ThrowTypeError("encountered radius too large to fit in unsigned");
+            }
+            baton->radius = static_cast<unsigned>(_radius);
+        } else {
+            baton->radius = 40;
```
Comment (yhahn) [inline src/binding.cpp]: If unspecified defaults to a radius of 40 miles like we do now.

File: src/binding.cpp [hunk 3]
```diff
@@ -1255,9 +1255,9 @@ inline bool contextSortByRelev(Context const& a, Context const& b) noexcept {
 }
 
 inline double tileDist(unsigned px, unsigned py, unsigned tileX, unsigned tileY) {
-    const double dx = static_cast<double>(px - tileX);
-    const double dy = static_cast<double>(py - tileY);
-    const double distance = dx * dx + dy * dy;
+    const double dx = static_cast<double>(px) - static_cast<double>(tileX);
+    const double dy = static_cast<double>(py) - static_cast<double>(tileY);
+    const double distance = sqrt((dx * dx) + (dy * dy));
```
Comment (apendleton) [inline src/binding.cpp]: I think the motivation for not taking the square root before was that it doesn't make a difference from an ordering perspective (if `sqrt(dx_1^2 + dy_1^2) > sqrt(dx_2^2 + dy_2^2)` it should also be true that `dx_1^2 + dy_1^2 > dx_2^2 + dy_2^2`) and calculating square roots is really slow. I guess now that we're doing real math with the results, though, it matters?

Comment (yhahn) [reply src/binding.cpp]: Yes, theoretically the calculated scoredist isn't just a sort value but can be lower than the 0-7 score of the feature and thus lose out, similar to the JS version in carmen.

I don't have a full grasp of whether this actually happens in practice yet though. For now, rationale in my head was to prepare for implementing the full approximate version of what happens in `carmen` in `carmen-cache` here.

**Pull request merged**